nat-lab/tests/test_batching.py

```diff
@@ -0,0 +1,304 @@
+# Batching works by checking if certain packets can be sent together. This means that
+# in order to test it's effects we can filter outgoing data and then check the time deltas.
+# When batching is in action time deltas should fall into the left side of the histogram because
+# deltas should be very small as packets leave in succession. There should be no action in
+# the middle and then histogram should have data on the interval boundary. Basically letter "U" shaped.
+# When using no batching, histogram might look very differently depending on the signal misalignment. Testing
+# depends highly on misalignment and must be taken into an account.
```

Comment:
Could you share more of your concerns? I sit hard on this problem and thought that actually using histograms for testing is a good way to test. Deadlines should be adhered to stricly in SessionKeeper, thus we can always expect a U shape with proper filtering.

In the current setup I don't see much possibilities for flakyness, unless of course, the testcae will do a `Ping()` then it will ruin it, but that's expected :)

Reply:
For example adding a sleep of 2s doesn't guarantee that the timers will be misaligned. They can just as well align at the 4s if the startup of postponed peer takes long enough. Unless I'm misunderstanding the test design :)

Reply:
You're right that if the meshnet peer registration would take too long, it has a possibility to align with the 4seconds, introducing a flake, however should that happen in NatLab? We control all the necessary infrastructure so that should be pretty fast I suppose?

I initially played with low numbers such as 1seconds for intervals, thus 500ms for sleeping and those were flaking a bit thus I increased by 4x. Since then locally, at least, it didn't flake through a number of times.

---

I think the numbers are pretty safe to go and we can just adjust them as the time goes if they flake.

I have an idea how to stremaline the flakyness a bit - I could introduce some constants instead of hardcoded numbers so in a case of a flake, it would be very easy to update, how does that sound?

Reply:
I do know that machines we use for CI can get veeeery slow, especially when disk I/O is taking place. I think that last year in December we observed lana tests failures that were the result of 2 orders of magnitude (if not more, the slack thread is obviously long gone so I can't verify this) slowdown of disk access compared to healthy runner. That runner got replaced but the issue can happen is less pronounced way but still severe enough to add few seconds of slowdown.
But as I said I don't really have a better idea, so maybe this is just something to keep in mind if those tests ever become flaky.